zebra-state/src/service/non_finalized_state.rs

```diff
@@ -86,6 +87,34 @@ impl NonFinalizedState {
             && self.network == other.network
     }
 
+    /// Returns an iterator over the non-finalized chains, with the best chain first.
+    //
+    // TODO: replace chain_set.iter().rev() with this method
+    pub fn chain_iter(&self) -> impl Iterator<Item = &Arc<Chain>> {
+        self.chain_set.iter()
+    }
+
+    /// Insert `chain` into `self.chain_set`, apply `chain_filter` to the chains,
+    /// then limit the number of tracked chains.
+    fn insert_with<F>(&mut self, chain: Arc<Chain>, chain_filter: F)
```

Comment:
Optional: Would an insert method that returns a drop guard for pruning the chain set be nice here?

```rust
struct PruneGaurd<'a>(&'a mut BTreeSet<Arc<Chain>>);

impl<'a> PruneGaurd<'a> {
    fn apply_and_prune<F: FnOnce(&mut BTreeSet<Arc<Chain>>)>(self, chain_filter: F) {
        chain_filter(self.0);
    }
}

impl<'a> Drop for PruneGaurd<'a> {
    fn drop(&mut self) {
        while self.0.len() > MAX_NON_FINALIZED_CHAIN_FORKS {
            // The first chain is the chain with the lowest work.
            self.0.pop_first();
        }
    }
}

/// Insert `chain` into `self.chain_set`, then limit the number of tracked chains.
fn insert(&mut self, chain: Arc<Chain>) -> PruneGaurd {
    self.chain_set.insert(chain);

    PruneGaurd(&mut self.chain_set)
}

pub fn commit_block(..) -> Result<(), ValidateContextError> {
    ..
    let prune_gaurd = self.insert(modified_chain);
    prune_gaurd.apply_and_prune(|chain_set| {
        chain_set.retain(|chain| chain.non_finalized_tip_hash() != parent_hash)
    });
    ..
}
```

Reply:
Drop guards aren't guaranteed to run, so it would be a less reliable implementation. What you're looking for is called a "linear type", which isn't implemented in Rust yet:
https://blog.yoshuawuyts.com/linearity-and-control/#what-are-linear-types

This particular implementation would also hold a mutable reference to the chains for the entire calling method, which would likely cause problems with the borrow-checker.